src/dir_status.ml

```diff
@@ -42,86 +47,76 @@ let check_no_module_consumer stanzas =
-          | None ->
-            if Path.equal dir project_root ||
-               is_standalone (get db ~dir:(Path.parent_exn dir)) then
-              Standalone (Some (ft_dir, None))
-            else
-              Is_component_of_a_group_but_not_the_root None
-          | Some d ->
-            match get_include_subdirs d.data with
-            | Some Unqualified ->
-              Group_root (ft_dir, d)
-            | Some No ->
-              Standalone (Some (ft_dir, Some d))
-            | None ->
-              if dir <> project_root &&
-                 not (is_standalone (get db ~dir:(Path.parent_exn dir)))
-              then begin
-                check_no_module_consumer d.data;
-                Is_component_of_a_group_but_not_the_root (Some d)
-              end else
-                Standalone (Some (ft_dir, Some d))
-      in
-      Hashtbl.add db.cache dir t;
-      t
-
-  let get_assuming_parent_is_part_of_group db ~dir ft_dir =
-    match Hashtbl.find db.cache (File_tree.Dir.path ft_dir) with
-    | Some t -> t
-    | None ->
-      let t =
-        match stanzas_in db ~dir with
-        | None -> Is_component_of_a_group_but_not_the_root None
-        | Some d ->
-          match get_include_subdirs d.data with
-          | Some Unqualified ->
-            Group_root (ft_dir, d)
-          | Some No ->
-            Standalone (Some (ft_dir, Some d))
-          | None ->
+  let get db ~dir =
+    let get ~dir = Memo.exec db.fn dir in
+    match
+      Option.bind (Path.drop_build_context dir)
+        ~f:(File_tree.find_dir db.file_tree)
+    with
+    | None -> begin
+        match Path.parent dir with
+        | None -> Standalone None
+        | Some dir ->
+          if is_standalone (get ~dir) then
+            Standalone None
+          else
+            Is_component_of_a_group_but_not_the_root None
+      end
+    | Some ft_dir ->
+      let project_root =
+        File_tree.Dir.project ft_dir
+        |> Dune_project.root
+        |> Path.of_local in
+      match stanzas_in db ~dir with
+      | None ->
+        if Path.equal dir project_root ||
+           is_standalone (get ~dir:(Path.parent_exn dir)) then
+          Standalone (Some (ft_dir, None))
+        else
+          Is_component_of_a_group_but_not_the_root None
+      | Some d ->
+        match get_include_subdirs d.data with
+        | Some Unqualified ->
+          Group_root (ft_dir, d)
+        | Some No ->
+          Standalone (Some (ft_dir, Some d))
+        | None ->
+          if dir <> project_root &&
+             not (is_standalone (get ~dir:(Path.parent_exn dir)))
+          then begin
             check_no_module_consumer d.data;
             Is_component_of_a_group_but_not_the_root (Some d)
-      in
-      Hashtbl.add db.cache dir t;
-      t
+          end else
+            Standalone (Some (ft_dir, Some d))
+
+  let make file_tree ~stanzas_per_dir =
+    let t =
+      { file_tree
+      ; stanzas_per_dir
+      ; fn =
+          Memo.create
```

Comment:
Note for later: it doesn't matter too much now, but eventually I was thinking of forbidding to memoize functions anywhere else than the toplevel of modules. Otherwise memoized functions might capture computed values that the system doesn't know about.

We'll have to refactor things a bit before we can do that here though.

Reply:
We discussed this and had two thoughts:

1. Maybe a memoized function created inside another function is all right (the dependency tracking seems capable of dealing with that). That said, "public" memoization and across-runs memoization should still only be made on top level.

2. The notion of "top level" doesn't have to be so strict: it's enough to forbid creation of the functions after some runtime initialization phase is finished. For now we don't want to work through details, but it's plausible that this code can stay this way.